geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java

```diff
@@ -0,0 +1,525 @@
+    @SuppressWarnings("rawtypes")
+    PartitionAttributes attributes = new PartitionAttributesFactory<String, String>()
+        .setTotalNumBuckets(BUCKETS)
+        .setLocalMaxMemory(0)
+        .create();
+
+    cacheRule.getCache().createRegionFactory(regionShortcut)
+        .setPartitionAttributes(attributes)
+        .create(REGION_NAME);
+
+  }
+
+  private void initDataStore(RegionShortcut regionShortcut) {
+    @SuppressWarnings("rawtypes")
+    PartitionAttributes attributes = new PartitionAttributesFactory<String, String>()
+        .setTotalNumBuckets(BUCKETS)
+        .create();
+
+    cacheRule.getCache().createRegionFactory(regionShortcut)
+        .setPartitionAttributes(attributes)
+        .create(REGION_NAME);
+  }
+
+  private void parametrizedSetup(RegionShortcut regionShortcut) {
+    server1.invoke(() -> initDataStore(regionShortcut));
+    server2.invoke(() -> initDataStore(regionShortcut));
+    accessor.invoke(() -> initAccessor(regionShortcut));
+  }
+
+  /**
+   * Waits until there are not PR messages in flight for the given region.
+   */
+  private void waitForSilence() {
+    PartitionedRegion region = (PartitionedRegion) cacheRule.getCache().getRegion(REGION_NAME);
+
+    await().untilAsserted(() -> {
+      PartitionedRegionStats prStats = region.getPrStats();
+      assertThat(prStats.getVolunteeringInProgress()).isEqualTo(0);
+      assertThat(prStats.getBucketCreatesInProgress()).isEqualTo(0);
+      assertThat(prStats.getPrimaryTransfersInProgress()).isEqualTo(0);
+      assertThat(prStats.getRebalanceBucketCreatesInProgress()).isEqualTo(0);
+      assertThat(prStats.getRebalancePrimaryTransfersInProgress()).isEqualTo(0);
+    });
+  }
+
+  private void assertRegionData(int entryCount, List<VM> vms) {
+    vms.forEach(vm -> vm.invoke(() -> {
+      waitForSilence();
+      Region<String, String> region = cacheRule.getCache().getRegion(REGION_NAME);
+
+      IntStream.range(0, entryCount)
+          .forEach(i -> assertThat(region.get(String.valueOf(i))).isEqualTo("Value_" + i));
+    }));
+  }
+
+  /**
+   * Asserts that the region is empty on requested VMs.
+   */
+  private void assertRegionIsEmpty(List<VM> vms) {
+    vms.forEach(vm -> vm.invoke(() -> {
+      waitForSilence();
+      Region<String, String> region = cacheRule.getCache().getRegion(REGION_NAME);
+
+      assertThat(region.isEmpty()).isTrue();
+    }));
+  }
+
+  /**
+   * Waits until there are no partition messages in flight and gets a snapshot of the region.
+   */
+  private Map<String, String> waitForSilenceAndGetRegionSnapshot() {
+    waitForSilence();
+    Map<String, String> regionSnapshot = new HashMap<>();
+    Region<String, String> region = cacheRule.getCache().getRegion(REGION_NAME);
+    region.forEach(regionSnapshot::put);
+
+    return regionSnapshot;
+  }
+
+  /**
+   * Asserts that the region data is consistent on requested VMs.
+   */
+  private void assertRegionDataConsistency(List<VM> vms) {
```

Comment:
This is going over the PRs entry/key set getting all the values. This will be consistent where-ever this is called. 
Are we trying to see if the data in one vm is same as in another vm. Or are we trying to see data in primary buckets are same as in secondary buckets.
Fetching the local PR data-set and iterating over it will give the data stored in that vm. 
Or just to see an op could be executed successfully after clear?

Reply:
The method asserts that the data is consistent across VMs by tacking a snapshot of the region per VM and comparing it against the other running VMs, it doesn't check primaries vs secondaries.

```
    vms.forEach(vm -> vm.invoke(() -> {
      final Map<String, String> thisSnapshot = waitForSilenceAndGetRegionSnapshot();
      assertThat(thisSnapshot).isEqualTo(vm0Snapshot);
      thisSnapshot.forEach((key, value) -> assertThat(value).isEqualTo("Value_" + key));
    }));

```

Reply:
What I am trying to says is, since you are iterating over the PR region; it will get the data from all the nodes. Thats remains same across all the nodes.

Reply:
@agingade : got your point, thanks for catching this. Will change the method to verify that the data both on the primary and secondary buckets are in sync.